drivers/flash/flash_rpi_pico.c

```diff
@@ -0,0 +1,318 @@
+			ssi->dr0 = (uint32_t) (tx ? *tx++ : 0);
+			--tx_count;
+			did_something = true;
+		}
+		if (rx_level) {
+			rxbyte = ssi->dr0;
+			did_something = true;
+			if (rx_skip) {
+				--rx_skip;
+			} else {
+				if (rx)
+					*rx++ = rxbyte;
+				--rx_count;
+			}
+		}
+
+		if (!did_something && __builtin_expect(flash_was_aborted(), 0))
+			break;
+	}
+	flash_cs_force(OUTOVER_HIGH);
+}
+
+void __no_inline_not_in_flash_func(flash_put_get_wrapper)(uint8_t cmd, const uint8_t *tx,
+					uint8_t *rx, size_t count)
+{
+	flash_cs_force(OUTOVER_LOW);
+	ssi->dr0 = cmd;
+	flash_put_get(tx, rx, count, 1);
+}
+
+static inline void flash_put_cmd_addr(uint8_t cmd, uint32_t addr)
+{
+	flash_cs_force(OUTOVER_LOW);
+	addr |= cmd << 24;
+	for (int i = 0; i < 4; ++i) {
+		ssi->dr0 = addr >> 24;
+		addr <<= 8;
+	}
+}
+
+void __no_inline_not_in_flash_func(flash_write_partial_internal)(uint32_t addr, const uint8_t *data,
+					size_t size)
+{
+	uint8_t status_reg;
+
+	flash_put_get_wrapper(FLASHCMD_WRITE_ENABLE, NULL, NULL, 0);
+	flash_put_cmd_addr(FLASHCMD_PAGE_PROGRAM, addr);
+	flash_put_get(data, NULL, size, 4);
+
+	do {
+		flash_put_get_wrapper(FLASHCMD_READ_STATUS, NULL, &status_reg, 1);
+	} while (status_reg & 0x1 && !flash_was_aborted());
+}
+
+void __no_inline_not_in_flash_func(flash_write_partial)(uint32_t flash_offs, const uint8_t *data,
+							size_t count)
+{
+	rom_connect_internal_flash_fn connect_internal_flash = (rom_connect_internal_flash_fn)
+					rom_func_lookup_inline(ROM_FUNC_CONNECT_INTERNAL_FLASH);
+	rom_flash_exit_xip_fn flash_exit_xip = (rom_flash_exit_xip_fn)
+						rom_func_lookup_inline(ROM_FUNC_FLASH_EXIT_XIP);
+	rom_flash_flush_cache_fn flash_flush_cache = (rom_flash_flush_cache_fn)
+						rom_func_lookup_inline(ROM_FUNC_FLASH_FLUSH_CACHE);
+
+	flash_init_boot2_copyout();
+
+	__compiler_memory_barrier();
+
+	connect_internal_flash();
+	flash_exit_xip();
+	flash_write_partial_internal(flash_offs, data, count);
+	flash_flush_cache();
+	flash_enable_xip_via_boot2();
+}
+
+static bool is_valid_range(off_t offset, uint32_t size)
+{
+	return (offset >= 0) && ((offset + size) <= FLASH_SIZE);
+}
+
+static int flash_rpi_read(const struct device *dev, off_t offset, void *data, size_t size)
+{
+	if (size == 0) {
+		return 0;
+	}
+
+	if (!is_valid_range(offset, size)) {
+		LOG_ERR("Read range exceeds the flash boundaries");
+		return -EINVAL;
+	}
+
+	memcpy(data, (uint8_t *)(CONFIG_FLASH_BASE_ADDRESS + offset), size);
+
+	return 0;
+}
+
+static int flash_rpi_write(const struct device *dev, off_t offset, const void *data, size_t size)
+{
+	uint32_t key;
+	size_t bytes_to_write;
+	uint8_t *data_pointer = (uint8_t *)data;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	if (!is_valid_range(offset, size)) {
+		LOG_ERR("Write range exceeds the flash boundaries. Offset=%#lx, Size=%u",
+				offset, size);
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	if (offset % PAGE_SIZE > 0) {
+		bytes_to_write = MIN(PAGE_SIZE - (offset % PAGE_SIZE), size);
```

Comment:
Assuming the PAGE_SIZE is power of 2, you can use binary operators here `PAGE_SIZE - (offset & (PAGE_SIZE -1))`, you did similar thing in line 231